src/nupic/types/Sdr.cpp

```diff
@@ -330,6 +330,22 @@ namespace sdr {
     }
 
 
+    void SparseDistributedRepresentation::killCells(Real fraction, UInt seed) {
+        NTA_CHECK( fraction >= 0.0 );
+        NTA_CHECK( fraction <= 1.0 );
+        int nkill = round( size * fraction );
+        Random rng(seed);
+        auto &data = getDense();
+        for( int i = size - 1; i >= 0; --i ) {
+            if( rng.getReal64() <= (Real64) nkill / (i+1) ) {
```

Comment:
this seems a complicated way of doing things(?) 
We want to kill nkill cells, right? So why don't use something as `Random::sample(n, from(=0), to(=size-1))` and then set all those samples to 0?

Reply:
`Random::sample(n, from(=0), to(=size-1))`

Because our C++ `Random::sample` does not accept a range.  I would have to build a list of every index in the SDR.

Alas, this isn't python where the random sample method accepts & is optimized for a `range` argument.

Reply:
> Because our C++ Random::sample does not accept a range. I would have to build a list of every index in the SDR.

I'd still say it's more readable? 
```
vector<UInt> indices;
for(...i<size) indices.push_back(i);
disableIndices = Random::sample(indices, nkill);
for( const auto dis : disableIndices) data[dis] = 0;
```

I don't clearly understand what the code `if( rng.getReal64() <= (Real64) nkill / (i+1) ) {` with changing the probability is doing?

Reply:
It's doing a random sample.  The probability that a cell is selected is equal to the number of cells left to be selected, divided by the total number of cells left.

Reply:
ok, might add just a comment //random sample; to that line

Reply:
Ok, I added an inline comment explaining this.  I think that now this is ready to merge.

Reply:
@ctrl-z-9000-times I still wasn't sure about uniform distribution with that implementation, so I changed to that of comment https://github.com/htm-community/nupic.cpp/pull/486#discussion_r288617332
Hope you don't mind, otherwise this is OK now :+1: